### PR TITLE
grt: mark FrNet as deleted when deleting a GrNet

### PR DESCRIPTION
**Pull request by arthurjolo**
Fixes [#4268](https://github.com/The-OpenROAD-Project/OpenROAD/issues/4268) 

Currently when a dbNet is deleted nothing is done to the FrNet that references it, so fastroute thinks that this net still exists. In some fastrout tries to access the dbNet that has been deleted causing a segfault. This PR solves this issue by marking a FrNet as deleted when it's dbNet is deleted, also it makes fastroute use only FrNets that are not marked as deleted.

### CODE REVIEW

**Comment by arthurjolo** [comment]
This PR has been throw a secure-CI, it is passing for the public designs.

File: src/grt/src/fastroute/src/maze.cpp [hunk 1]
```diff
@@ -1382,6 +1382,9 @@ void FastRouteCore::mazeRouteMSMD(const int iter,
     if (nets_[netID]->isRouted())
       continue;
 
+    if (nets_[netID]->isDeleted())
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/maze.cpp:1385]: Ditto

File: src/grt/src/fastroute/src/route.cpp [hunk 2]
```diff
@@ -233,6 +233,9 @@ void FastRouteCore::routeLAll(bool firstTime)
       if (nets_[i]->isRouted())
         continue;
 
+      if (nets_[i]->isDeleted())
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/route.cpp:236]: Merge `ifs`

File: src/grt/src/fastroute/src/route.cpp [hunk 3]
```diff
@@ -242,6 +245,9 @@ void FastRouteCore::routeLAll(bool firstTime)
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/route.cpp:248]: Ditto

File: src/grt/src/fastroute/src/route.cpp [hunk 4]
```diff
@@ -253,6 +259,9 @@ void FastRouteCore::routeLAll(bool firstTime)
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/route.cpp:262]: Ditto

File: src/grt/src/fastroute/src/route.cpp [hunk 5]
```diff
@@ -1062,6 +1071,9 @@ void FastRouteCore::routeMonotonicAll(int threshold)
     if (nets_[netID]->isRouted())
       continue;
 
+    if (nets_[netID]->isDeleted())
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/route.cpp:1074]: Ditto

File: src/grt/src/fastroute/src/route.cpp [hunk 6]
```diff
@@ -1247,6 +1259,9 @@ void FastRouteCore::spiralRouteAll()
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/route.cpp:1262]: Ditto

File: src/grt/src/fastroute/src/route.cpp [hunk 7]
```diff
@@ -1297,6 +1312,9 @@ void FastRouteCore::spiralRouteAll()
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/route.cpp:1315]: Ditto

File: src/grt/src/fastroute/src/route.cpp [hunk 8]
```diff
@@ -1326,6 +1344,9 @@ void FastRouteCore::spiralRouteAll()
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/route.cpp:1347]: Ditto

File: src/grt/src/fastroute/src/route.cpp [hunk 9]
```diff
@@ -1381,6 +1402,9 @@ void FastRouteCore::spiralRouteAll()
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/route.cpp:1405]: Ditto

File: src/grt/src/fastroute/src/route.cpp [hunk 10]
```diff
@@ -1723,6 +1747,9 @@ void FastRouteCore::routeLVAll(int threshold, int expand, float logis_cof)
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/route.cpp:1750]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 11]
```diff
@@ -70,6 +70,9 @@ void FastRouteCore::ConvertToFull3DType2()
     if (nets_[netID]->isRouted())
       continue;
 
+    if (nets_[netID]->isDeleted())
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:73]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 12]
```diff
@@ -149,6 +152,9 @@ void FastRouteCore::netpinOrderInc()
     if (nets_[j]->isRouted())
       continue;
 
+    if (nets_[j]->isDeleted())
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:155]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 13]
```diff
@@ -180,7 +186,8 @@ void FastRouteCore::fillVIA()
   for (int netID = 0; netID < netCount(); netID++) {
     if (nets_[netID]->isRouted())
       continue;
-
+    if (nets_[netID]->isDeleted())
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:189]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 14]
```diff
@@ -324,6 +331,11 @@ int FastRouteCore::threeDVIA()
     if (nets_[netID]->isRouted()) {
       continue;
     }
+
+    if (nets_[netID]->isDeleted()) {
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:335]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 15]
```diff
@@ -753,6 +765,9 @@ void FastRouteCore::layerAssignmentV4()
     if (nets_[netID]->isRouted())
       continue;
 
+    if (nets_[netID]->isDeleted())
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:768]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 16]
```diff
@@ -771,6 +786,9 @@ void FastRouteCore::layerAssignmentV4()
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:789]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 17]
```diff
@@ -893,6 +911,9 @@ void FastRouteCore::layerAssignment()
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:914]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 18]
```diff
@@ -943,6 +964,9 @@ void FastRouteCore::layerAssignment()
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:967]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 19]
```diff
@@ -1026,6 +1050,9 @@ void FastRouteCore::checkRoute3D()
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:1053]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 20]
```diff
@@ -1151,6 +1178,9 @@ void FastRouteCore::StNetOrder()
     if (nets_[j]->isRouted())
       continue;
 
+    if (nets_[j]->isDeleted())
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:1181]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 21]
```diff
@@ -1218,6 +1248,9 @@ float FastRouteCore::CalculatePartialSlack()
     if (nets_[netID]->isRouted()) {
       continue;
     }
+    if (nets_[netID]->isDeleted()) {
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:1251]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 22]
```diff
@@ -1239,6 +1272,9 @@ float FastRouteCore::CalculatePartialSlack()
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:1275]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 23]
```diff
@@ -1330,6 +1366,10 @@ void FastRouteCore::removeLoops()
       continue;
     }
 
+    if (nets_[netID]->isDeleted()) {
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:1369]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 24]
```diff
@@ -1753,6 +1799,9 @@ void FastRouteCore::copyRS(void)
       if (nets_[netID]->isRouted())
         continue;
 
+      if (nets_[netID]->isDeleted())
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:1802]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 25]
```diff
@@ -1770,6 +1819,9 @@ void FastRouteCore::copyRS(void)
     if (nets_[netID]->isRouted())
       continue;
 
+    if (nets_[netID]->isDeleted())
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:1822]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 26]
```diff
@@ -1824,6 +1876,9 @@ void FastRouteCore::copyBR(void)
       if (nets_[netID]->isRouted())
         continue;
 
+      if (nets_[netID]->isDeleted())
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:1879]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 27]
```diff
@@ -1837,6 +1892,9 @@ void FastRouteCore::copyBR(void)
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:1895]: Ditto

File: src/grt/src/fastroute/src/utility.cpp [hunk 28]
```diff
@@ -1936,6 +1994,9 @@ void FastRouteCore::freeRR(void)
...
```
Comment (eder-matheus) [inline src/grt/src/fastroute/src/utility.cpp:1997]: Ditto

File: src/grt/src/fastroute/src/FastRoute.cpp [hunk 29]
```diff
@@ -283,6 +283,12 @@ FrNet* FastRouteCore::addNet(odb::dbNet* db_net,
   return net;
 }
 
+void FastRouteCore::removeNet(odb::dbNet* db_net)
+{
+  int netID = db_net_id_map_[db_net];
+  nets_[netID]->setIsDeleted(true);
```
[the remark this answers is not in the document]
Comment (eder-matheus) [reply src/grt/src/fastroute/src/FastRoute.cpp:289]: @arthurjolo just this missing request.

File: src/grt/src/fastroute/src/FastRoute.cpp [hunk 30]
```diff
@@ -727,6 +735,9 @@ NetRouteMap FastRouteCore::getPlanarRoutes()
     if (nets_[netID]->isRouted()) {
       continue;
     }
+    if (nets_[netID]->isDeleted()) {
```
[the remark this answers is not in the document]
Comment (eder-matheus) [reply src/grt/src/fastroute/src/FastRoute.cpp:737]: @arthurjolo And here too. If you already added, maybe this is a GitHub issue.

**Comment by maliberty** [comment]
What is the reason to flag it as deleted vs just deleting it?

**Comment by arthurjolo** [comment]
> What is the reason to flag it as deleted vs just deleting it?

A flag is used because fastroute is heavly deppendend on the order of the nets on the vector _nets__. Talking with eder we decided to use the flag as a temporary solution and unlock the user from the error. And keep working on a solution to correctly delete the nets.

File: src/grt/src/fastroute/src/FastRoute.cpp [hunk 31]
```diff
@@ -724,7 +731,7 @@ NetRouteMap FastRouteCore::getPlanarRoutes()
   // Get routes before layer assignment
 
   for (int netID = 0; netID < netCount(); netID++) {
-    if (nets_[netID]->isRouted()) {
+    if (nets_[netID]->isRouted() || nets_[netID]->isDeleted()) {
```
[the remark this answers is not in the document]
Comment (arthurjolo) [reply src/grt/src/fastroute/src/FastRoute.cpp:736]: I am not sure what is the sugestion here?

[the remark this answers is not in the document]
Comment (eder-matheus) [reply src/grt/src/fastroute/src/FastRoute.cpp:736]: Looks like a clang-tidy bug? You can ignore this one.

**Review by eder-matheus — APPROVED** [review]

**Pull request merged**